src/test/java/io/cdap/plugin/successfactors/source/SuccessFactorsSourceTest.java

```diff
@@ -0,0 +1,467 @@
+      .password("password")
+      .selectOption("col1,col2,   \n  parent/col1,\r       col3     ");
+    try {
+      pluginConfig = pluginConfigBuilder.build();
+      successFactorsSource = new SuccessFactorsSource(pluginConfig);
+      Map<String, Object> plugins = new HashMap<>();
+      MockPipelineConfigurer mockPipelineConfigurer = new MockPipelineConfigurer(null, plugins);
+      successFactorsSource.configurePipeline(mockPipelineConfigurer);
+      Assert.fail("Exception is not thrown if valid reference name is provided");
+    } catch (ValidationException ve) {
+      List<ValidationFailure> failures = ve.getFailures();
+      Assert.assertEquals(1, failures.size());
+      Assert.assertEquals("Invalid reference name ''.", ve.getFailures().get(0).getMessage());
+    }
+  }
+
+  private URL getUrl() throws MalformedURLException {
+    return new URL("http://localhost:8000/odata/v2/entity-name");
+  }
+
+  private SuccessFactorsResponseContainer getResponseContainer() {
+    return new SuccessFactorsResponseContainer(200, "ok",
+                                               "2.0", new byte[]{50});
+  }
+
+  private Schema getPluginSchema() throws IOException {
+    String schemaString = "{\"type\":\"record\",\"name\":\"SuccessFactorsColumnMetadata\",\"fields\":[{\"name\":" +
+      "\"backgroundElementId\",\"type\":\"long\"},{\"name\":\"bgOrderPos\",\"type\":\"long\"},{\"name\":" +
+      "\"description\",\"type\":[\"string\",\"null\"]},{\"name\":\"endDate\",\"type\":[{\"type\":\"long\"," +
+      "\"logicalType\":\"timestamp-micros\"},\"null\"]},{\"name\":\"lastModifiedDate\",\"type\":" +
+      "[{\"type\":\"long\",\"logicalType\":\"timestamp-micros\"},\"null\"]},{\"name\":\"project\",\"type\":" +
+      "\"string\"},{\"name\":\"startDate\",\"type\":[{\"type\":\"long\",\"logicalType\":\"timestamp-micros\"}," +
+      "\"null\"]},{\"name\":\"userId\",\"type\":\"string\"}]}";
+
+    return Schema.parseJson(schemaString);
+  }
+
+  @Test
+  public void testConfigurePipelineWSchemaNotNull() throws SuccessFactorsServiceException, TransportException,
+    IOException {
+    pluginConfigBuilder = SuccessFactorsPluginConfig.builder()
+      .referenceName("unit-test-ref-name")
+      .baseURL("http://localhost")
+      .entityName("entity-name")
+      .username("username")
+      .password("password");
+
+    pluginConfig = pluginConfigBuilder.build();
+    successFactorsTransporter = new SuccessFactorsTransporter(pluginConfig.getUsername(), pluginConfig.getPassword());
+    successFactorsUrlContainer = new SuccessFactorsUrlContainer(pluginConfig);
+    successFactorsSchemaGenerator = new SuccessFactorsSchemaGenerator(new SuccessFactorsEntityProvider(edm));
+
+    new Expectations(SuccessFactorsUrlContainer.class, SuccessFactorsTransporter.class,
+                     SuccessFactorsSchemaGenerator.class) {
+      {
+        successFactorsUrlContainer.getTesterURL();
+        result = getUrl();
+        minTimes = 0;
```

Comment:
instead of setting minTimes = 0 in these tests can we set times = 1 (or whatever the expected number of invocations is? )

Reply:
Done.